arboretum/kubernetes/fetchers/fetch_cluster_resource.py

```diff
@@ -0,0 +1,159 @@
+        resources = {}
+        for resource_type in resource_types:
+            if '/' in resource_type:
+                url = f'{server_url}/apis/{resource_type}'
+            else:
+                url = f'{server_url}/api/v1/{resource_type}'
+            resp = s.get(url, verify=verify)
+            if resp.status_code == 404:
+                self.logger.warning(
+                    'Cluster "%s" does not provide resource type "%s"',
+                    server_url,
+                    resource_type
+                )
+                continue
+            resp.raise_for_status()
+            resources[resource_type] = resp.json()['items']
+        return resources
+
+    @store_raw_evidence('kubernetes/cluster_resource.json')
+    def fetch_cluster_resource(self):
+        """Fetch cluster resources of listed clusters."""
+        ev = get_evidence_dependency(
+            'raw/auditree/compliance_config.json', self.locker
+        )
+        config = json.loads(ev.content)['org']['kubernetes']
+
+        cluster_list = config['cluster_list']
+        self.logger.debug('number of clusters: %d', len(cluster_list))
+
+        resource_types = self.RESOURCE_TYPE_DEFAULT
+        if ('cluster_resource' in config
+                and 'target_resource_types' in config['cluster_resource']):
+            resource_types = (
+                config['cluster_resource']['target_resource_types']
+            )
+        resources = {}
+
+        for c in cluster_list:
+            self.logger.info('Getting resource from cluster: %s', c['name'])
+            try:
+                with open(c['kubeconfig']) as f:
+                    kubeconfig = yaml.safe_load(f)
+            except OSError:
+                self.logger.error(
+                    'Cannot access cluster "%s" because '
+                    'kubeconfig file is not found: %s',
+                    c['serverURL'],
+                    c['kubeconfig']
+                )
+                raise
+            for cluster in kubeconfig['clusters']:
+                ca_filepath = cluster['cluster']['certificate-authority']
+                server_url = cluster['cluster']['server']
+                cluster_name = cluster['name']
+                if cluster_name != c['name']:
+                    self.logger.info(
+                        'Skip non-target cluster: %s', cluster_name
+                    )
+                    continue
+
+                # look up username
+                username = None
+                for context in kubeconfig['contexts']:
+                    if context['context']['cluster'] == cluster_name:
+                        username = context['context']['user']
+                        break
+                if username is None:
+                    raise RuntimeError(
+                        'Failed to get username for cluster '
+                        f'{cluster["name"]} (server: '
+                        f'{cluster["cluster"]["server"]})'
+                    )
```

Comment:
I'd remove this and let Python raise the error when the `username` is actually referenced.

Reply:
I will simply remove this `raise`; when `username == None`, token will also become `None`, and thus the error will be cached at https://github.com/ComplianceAsCode/auditree-arboretum/pull/32#discussion_r500499442